marshmallow-code/webargs
# Only allow users to specify a single location per parse call & pass full location data to schema.load

This is still incomplete, but core and flaskparser tests are now passing. No work is (yet) included on other concrete parsers, but I fixed the base asyncparser to make mypy linting pass.

Resolves #419, #267, #164, #268 
Obsoletes, and therefore closes, #410

I've updated the changelog with a basic "statement of intent". As I work on this, I will try to maintain it as an accurate record of how this change should look to the outside world. Obviously, narrative docs also need to be updated with relevant info.

Highlights / greatest hits:
- Marks this as an unreleased v6.0.0 in the changelog
- Replace `DEFAULT_LOCATIONS` with `DEFAULT_LOCATION="json"`
- Replace `parse_{location}` with `location_load_{location}`
- Replace `locations=<iterable>` with `location=<str>`
- Remove `locations=...` from Fields
- Remove `Parser.parse_arg`
- Remove `webargs.core.get_value`
- Add `webargs.multidictproxy.MultiDictProxy` which has `get_value`-like behavior and proxies various dict-like objects
- Reverse the decision made in #297 with [a comment in tests explaining this change](https://github.com/sirosen/webargs/blob/90c337278677aaca603b1b8f71aa914ceaad8dca/src/webargs/testing.py#L205-L213) and a bit of [similar explanation in the commit message](
https://github.com/marshmallow-code/webargs/commit/e59963f3994e26ee108226d52536af943ec9b961)
- Numerous fixes to tests to be more explicit about which location is being used (since you can't trivially have a single location `/echo` which loads data from a bunch of locations)
- As a side-effect of the above changes, certain inputs which were previously ignored are treated as errors (e.g. `1` for a JSON body will now parse and be passed literally to the schema, which will fail because `1` is not a valid input for `schema.load`). Changes to the tests serve as a somewhat-readable list of behavioral changes.

To try to keep this from sprawling *even more*, my intent is to address some items mentioned in #419 either in follow-up PRs or as part of this only if we're happy with this direction (and agree on those items). In particular: the `json_or_form` location I proposed and nested `use_args` usage and error collection.

@sloria, @lafrech, my biggest question is: do you like where this is going? Should I just hammer out all of the other parsers?
